### PR TITLE
Add support for FFI while leaving legacy implementations intact

### PR DESCRIPTION
**Pull request by JasonLunn**
Alternative to #10221

### CODE REVIEW

File: ruby/lib/google/protobuf/ffi/field_descriptor.rb [hunk 1]
```diff
@@ -0,0 +1,326 @@
...
+        instance.send(:initialize, field_def, descriptor_pool)
+        instance
+      end
+
+      # TODO(jatl) Can this be added to the public API?
+      def real_containing_oneof
+        @real_containing_oneof ||= Google::Protobuf::FFI.real_containing_oneof self
+      end
+
+      # Implementation details below are subject to breaking changes without
```
Comment (haberman) [inline ruby/lib/google/protobuf/ffi/field_descriptor.rb:268]: I don't think a comment will be enough to communicate this. Perhaps a name that starts with `__internal` or similar?

File: ruby/lib/google/protobuf/ffi/descriptor.rb [hunk 2]
```diff
@@ -0,0 +1,765 @@
...
+          message = descriptor.msgclass.send(:private_constructor, arena, msg: msg)
+        end
+        message
+      end
+
+      def pool
+        @descriptor_pool
+      end
+
+      def build_message_class
```
Comment (haberman) [inline ruby/lib/google/protobuf/ffi/descriptor.rb:176]: I think it would help readability if this method was in its own file called message.rb.

File: ruby/lib/google/protobuf/ffi/descriptor.rb [hunk 3]
```diff
@@ -0,0 +1,765 @@
...
+              end
+              if size >= buffer_size
+                raise ParseError.new "Inconsistent JSON encoding sizes - was #{buffer_size - 1}, now #{size}"
+              end
+            end
+
+            buffer.read_string_length(size).force_encoding("UTF-8").freeze
+          end
+
+          @descriptor.each do |field_descriptor|
```
Comment (haberman) [inline ruby/lib/google/protobuf/ffi/descriptor.rb:418]: Ditto, can we move this into its own function? `add_accessors_for_field` or similar?

File: ruby/lib/google/protobuf/ffi/descriptor.rb [hunk 4]
```diff
@@ -0,0 +1,765 @@
...
+                end
+              end
+              define_method("has_#{field_name}?") do
+                !Google::Protobuf::FFI.get_message_which_oneof(@msg, oneof_descriptor).nil?
+              end
+            end
+          end
+
+          private
+          # Implementation details below are subject to breaking changes without
```
Comment (haberman) [inline ruby/lib/google/protobuf/ffi/descriptor.rb:525]: If these are truly private, do we need this warning?

File: ruby/lib/google/protobuf/ffi/descriptor.rb [hunk 5]
```diff
@@ -0,0 +1,765 @@
...
+              end
+            end
+            "<#{name}: #{field_output.join(', ')}>"
+          end
+
+          ##
+          # Gets a field of this message identified by the argument definition.
+          #
+          # @param field [FieldDescriptor] Descriptor of the field to get
+          def get_field(field, unwrap: false)
```
Comment (haberman) [inline ruby/lib/google/protobuf/ffi/descriptor.rb:680]: The code currently switches back and forth between "static" methods `def self.foo()` and regular methods `def foo()`. This makes the code somewhat hard to follow. Could we have it organized so it mostly puts all the methods of one kind together?

**Review by haberman — CHANGES_REQUESTED** [review]

**Comment by jskeet** [comment]
I have no context for this change, and I suspect that a large number of the 1,447 commits are existing ones. Is there any reason not to *rebase* this branch over main, so the actual changes are clearer? (And does this actually do anything for C# at all? If not, I'll remove myself from the reviewer list...)

**Comment by JasonLunn** [comment]
> I have no context for this change, and I suspect that a large number of the 1,447 commits are existing ones. Is there any reason not to _rebase_ this branch over main, so the actual changes are clearer? (And does this actually do anything for C# at all? If not, I'll remove myself from the reviewer list...)

@jskeet - I have no idea how this branch picked up so much cruft. Please feel free to remove yourself as a reviewer, this PR is completely Ruby-centric.

File: php/README.md [hunk 6]
```diff
@@ -14,10 +14,14 @@ generation functionality.
 
 ## Requirements
 
-To use PHP runtime library requires:
+Using the PHP runtime library requires:
 
-- C extension: PHP 7.x, 8.0
-- [PHP package](http://php.net/downloads.php): PHP 5.5, 5.6, 7.x, or 8.0.
+-   C extension: PHP 7.x, 8.x
```
Comment (Logofile) [inline php/README.md:19]: In the news article that should be published today: "As per our official [PHP support policy](https://cloud.google.com/php/getting-started/supported-php-versions), we will be dropping support for PHP 7.4 and lower. This means the minimum supported PHP version is 8.0."

**Review by Logofile — COMMENTED** [review]
My review included only README and other Markdown files. LGTM.

File: ruby/lib/google/protobuf/ffi/field_descriptor.rb [hunk 7]
```diff
@@ -0,0 +1,326 @@
...
+        def from_native(field_def, _ = nil)
+          return nil if field_def.nil? or field_def.null?
+          # Calling upb_FieldDef_File(field_def) would create a cyclic
+          # dependency because either 1) we'd have to define the method to accept
+          # an untyped pointer or 2) FFI would complain about passing a
+          # FFI::Pointer instance instead of a FieldDescriptor. Instead, directly
+          # read the top of the FieldDef structure and extract the FileDef*.
+          field_def_struct = Google::Protobuf::FFI::Upb_FieldDef.new(field_def)
+          file_def = field_def_struct[:file_def]
+          raise RuntimeError.new "FileDef is nil" if file_def.nil?
```
[the remark this answers is not in the document]
Comment (JasonLunn) [reply ruby/lib/google/protobuf/ffi/field_descriptor.rb:67]: `nil?` is defined on every Ruby object. `null?` is a method only on the `FFI::Pointer` class hierarchy. While writing the original PR last year, I ran into both bugs that triggered both situations when I'd misconfigured one thing or another, and it was helpful to be able to differentiate between the two scenarios.

I don't feel strongly about having this level of runtime error checking now, as I never see any violations of these anymore. Let's chat offline about how much defense to play here.

File: ruby/lib/google/protobuf/ffi/field_descriptor.rb [hunk 8]
```diff
@@ -0,0 +1,326 @@
...
+
+      # Implementation details below are subject to breaking changes without
+      # warning and are intended for use only within the gem.
+
+      ##
+      # Sets the field this FieldDescriptor represents to the given value on the given message.
+      # @param value [Object] Value to be set
+      # @param msg [::FFI::Pointer] Pointer the the upb_Message
+      # @param arena [Arena] Arena of the message that owns msg
+      def set_value_on_message(value, msg, arena, wrap: false)
```
[the remark this answers is not in the document]
Comment (JasonLunn) [reply ruby/lib/google/protobuf/ffi/field_descriptor.rb:276]: All the methods below line 250 are `private`.

File: ruby/lib/google/protobuf/ffi/ffi.rb [hunk 9]
```diff
@@ -0,0 +1,371 @@
...
+        end
+      end
+
+      # Arena
+      attach_function :create_arena, :Arena_create,         [], Arena
+      attach_function :fuse_arena,   :upb_Arena_Fuse,       [Arena, Arena], :bool
+      # Argument takes a :pointer rather than a typed Arena here due to
+      # implementation details of FFI::AutoPointer.
+      attach_function :free_arena,   :upb_Arena_Free,       [:pointer], :void
+      attach_function :arena_malloc, :upb_Arena_FastMalloc, [Arena, :size_t], :pointer
```
[the remark this answers is not in the document]
Comment (JasonLunn) [reply ruby/lib/google/protobuf/ffi/ffi.rb:259]: Something was probably lost in translation along the way. Currently, `upb_Arena_FastMalloc` is (at least now?) a function that is defined in `glue.c` and just passes through a call to the real `upb_Arena_Malloc`, which already has the API visibility I need to call it directly. Should be straightforward to clean this up.

[the remark this answers is not in the document]
Comment (JasonLunn) [reply ruby/lib/google/protobuf/ffi/ffi.rb]: Done

File: ruby/lib/google/protobuf/ffi/descriptor.rb (hunk 2 above)
Comment (JasonLunn) [reply ruby/lib/google/protobuf/ffi/descriptor.rb:176]: Done.

File: ruby/lib/google/protobuf/ffi/descriptor.rb (hunk 4 above)
Comment (JasonLunn) [reply ruby/lib/google/protobuf/ffi/descriptor.rb:532]: I don't feel strongly about this, but since we'd gone to extraordinary measures in the CRuby implementation to frustrate direct calls from Ruby for various methods, I opted to be more explicit. If you want me to remove them from across the PR, let me know.

File: ruby/BUILD.bazel [hunk 10]
```diff
@@ -11,6 +13,29 @@ load("//conformance:defs.bzl", "conformance_test")
...
+    name = "ffi",
+    build_setting_default = "disabled",
+    values = [
+        "enabled",
+        "disabled",
+    ],
+)
+
+config_setting(
+    name = "ffi_enabled",
```
[the remark this answers is not in the document]
Comment (JasonLunn) [reply ruby/BUILD.bazel:28]: Can you point me at an example?

File: ruby/lib/google/protobuf/ffi/descriptor.rb (hunk 5 above)
Comment (JasonLunn) [reply ruby/lib/google/protobuf/ffi/descriptor.rb:687]: Done

File: ruby/lib/google/protobuf/ffi/descriptor.rb (hunk 3 above)
Comment (JasonLunn) [reply ruby/lib/google/protobuf/ffi/descriptor.rb:428]: Done

File: ruby/lib/google/protobuf/ffi/field_descriptor.rb (hunk 1 above)
Comment (JasonLunn) [reply ruby/lib/google/protobuf/ffi/field_descriptor.rb:268]: Move this up to next to the `private` keyword as it was originally intended.

File: ruby/lib/google/protobuf/ffi/field_descriptor.rb [hunk 11]
```diff
@@ -0,0 +1,326 @@
...
+        end
+
+        ##
+        # @param field_def [::FFI::Pointer] FieldDef pointer to be wrapped
+        # @param _ [Object] Unused
+        def from_native(field_def, _ = nil)
+          return nil if field_def.nil? or field_def.null?
+          # Calling upb_FieldDef_File(field_def) would create a cyclic
+          # dependency because either 1) we'd have to define the method to accept
+          # an untyped pointer or 2) FFI would complain about passing a
```
[the remark this answers is not in the document]
Comment (JasonLunn) [reply ruby/lib/google/protobuf/ffi/field_descriptor.rb:62]: Done. Removed all the FFI struct definitions and introduced functions that take untyped pointers instead.

File: ruby/lib/google/protobuf/ffi/field_descriptor.rb [hunk 12]
```diff
@@ -0,0 +1,326 @@
...
+          # FFI::Pointer instance instead of a FieldDescriptor. Instead, directly
+          # read the top of the FieldDef structure and extract the FileDef*.
+          field_def_struct = Google::Protobuf::FFI::Upb_FieldDef.new(field_def)
+          file_def = field_def_struct[:file_def]
+          raise RuntimeError.new "FileDef is nil" if file_def.nil?
+          raise RuntimeError.new "FileDef is null" if file_def.null?
+          pool_def = Google::Protobuf::FFI.file_def_pool file_def
+          raise RuntimeError.new "PoolDef is nil" if pool_def.nil?
```
[the remark this answers is not in the document]
Comment (JasonLunn) [reply ruby/lib/google/protobuf/ffi/field_descriptor.rb:70]: Done. See PointerHelper.

File: ruby/lib/google/protobuf/ffi/ffi.rb [hunk 13]
```diff
@@ -0,0 +1,371 @@
...
+               :msg, [:char, Upb_Status_MaxMessage]
+
+        def initialize
+          super
+          FFI.clear self
+        end
+      end
+
+      # Arena
+      attach_function :create_arena, :Arena_create,         [], Arena
```
[the remark this answers is not in the document]
Comment (JasonLunn) [reply ruby/lib/google/protobuf/ffi/ffi.rb:232]: `ext/google/protobuf_c/glue.c`

File: ruby/lib/google/protobuf/ffi/internal/convert.rb [hunk 14]
```diff
@@ -0,0 +1,342 @@
...
+            string_value = value.encode("ASCII-8BIT")
+            return_value[:str_val][:size] = string_value.bytesize
+            return_value[:str_val][:data] = Google::Protobuf::FFI.arena_malloc(arena, string_value.bytesize)
+            raise NoMemoryError.new "Cannot allocate #{string_value.bytesize} bytes for bytes on Arena" if return_value[:str_val][:data].nil? || return_value[:str_val][:data].null?
+            return_value[:str_val][:data].write_string_length(string_value, string_value.bytesize)
+          when :message
+            raise TypeError.new "nil message not allowed here." if value.nil?
+            if value.is_a? Hash
+              raise RuntimeError.new "Attempted to initialize message from Hash for field #{name} but have no definition" if msg_or_enum_def.nil?
+              new_message = msg_or_enum_def.msgclass.
```
[the remark this answers is not in the document]
Comment (JasonLunn) [reply ruby/lib/google/protobuf/ffi/internal/convert.rb:84]: `convert_ruby_to_upb` has more responsibilities than the linked code in `message.c`, because the way the calling code is factored is simpler and thus the cases in the current CRuby implementation where hash arguments are supported also run through here. To break hash support for this branch in a way that maintains 1:1 feature parity would require non-trivial rework.

File: ruby/lib/google/protobuf/ffi/descriptor.rb [hunk 15]
```diff
@@ -0,0 +1,765 @@
...
+            end
+          end
+
+          def self.encode_internal(msg, encoding_options = 0)
+            temporary_arena = Google::Protobuf::FFI.create_arena
+
+            mini_table_ptr = Google::Protobuf::FFI.get_mini_table(descriptor)
+            size_ptr = ::FFI::MemoryPointer.new(:size_t, 1)
+            encoding = Google::Protobuf::FFI.encode_message(msg, mini_table_ptr, encoding_options, temporary_arena, size_ptr)
+            yield encoding, size_ptr.read(:size_t), mini_table_ptr
```
[the remark this answers is not in the document]
Comment (JasonLunn) [reply ruby/lib/google/protobuf/ffi/descriptor.rb]: I can't recall - possibly made more sense in an earlier version, or possibly just a habit when there are multiple return values.

File: ruby/lib/google/protobuf/ffi/internal/convert.rb [hunk 16]
```diff
@@ -0,0 +1,342 @@
...
+              arena.fuse(value.instance_variable_get(:@arena))
+              return_value[:msg_val] = value.instance_variable_get :@msg
+            end
+          when :enum
+            return_value[:int32_val] = case value
+              when Numeric
+                value.to_i
+              when String, Symbol
+                enum_number = EnumDescriptor.send(:lookup_name, msg_or_enum_def, value.to_s)
+                #TODO(jatl) add the bad value to the error message after tests pass
```
[the remark this answers is not in the document]
Comment (JasonLunn) [reply ruby/lib/google/protobuf/ffi/internal/convert.rb:131]: ✅ Agreed

File: ruby/lib/google/protobuf/ffi/internal/convert.rb [hunk 17]
```diff
@@ -0,0 +1,342 @@
...
+            hash_key = convert_upb_to_ruby(key_message_value, key_field_type)
+            hash_value = scalar_create_hash(value_message_value, value_field_type, msg_or_enum_descriptor: value_field_def.subtype)
+            return_value[hash_key] = hash_value
+          end
+          return_value
+        end
+
+        def repeated_field_create_array(array, field_descriptor, type)
+          return_value = []
+          # puts "JATL - about to call FFI method array_size(#{array}); stacktrace:\n#{caller.join("\n")}"
```
[the remark this answers is not in the document]
Comment (JasonLunn) [reply ruby/lib/google/protobuf/ffi/internal/convert.rb:280]: Done

File: ruby/BUILD.bazel (hunk 10 above)
[the remark this answers is not in the document]
Comment (JasonLunn) [reply ruby/BUILD.bazel:28]: I will create the new test configurations in a separate PR from a branch within protocolbuffers so that CI will run it.
